httpx/utils.py

```diff
@@ -94,6 +95,7 @@ def guess_json_utf(data: bytes) -> typing.Optional[str]:
 NETRC_STATIC_FILES = (Path("~/.netrc"), Path("~/_netrc"))
 
 
+@lru_cache()
```

Comment:
Because of the `host` parameter, we're now still looking up NETRC logins once per host. Since `netrc_info` doesn't depend on `host` at all, how about moving all logic associated to looking it up in a separate function, and apply `@functools.lru_cache()` to *that* instead?

```python
import functools

@functools.lru_cache
def get_netrc():
    ...
    return netrc_info

def get_netrc_login(host: str) -> typing.Optional[typing.Tuple[str, str, str]]:
    netrc = get_netrc()  # This is only fetched on 1st call.
    return netrc.authenticators(host)  # type: ignore
```

Reply:
Does it make sense to use lru_cache if the values won't change?

Reply:
Well um, I guess the alternative is manually caching the result into a module-level variable:

```python
_NETRC = None

def get_netrc_login(host: str)
    nolocal _NETRC
    if _NETRC is None:
        _NETRC = get_netrc()
    return _NETRC.authenticators(host)
```

Pushing this idea further, what we could do instead is to do caching at the `Client` level: have a `self.netrc = None` attribute, populate it in `.get_auth_middleware()` if/when we need to, and then use `self.netrc.authenticators(host)` to get the logins.

That way, we don't magically cache anything in the utilities, and it makes everything simpler and more testable.

What do you think @cansarigol?

Reply:
I pushed a commit about this. Is it what you want? Actually I agree with you because of get_netrc_login is used just in get_auth_middleware. More clear for me too. if it is appropriate, I will fix the tests.